### PR TITLE
elasticlunr: add type definitions

### PR DESCRIPTION
**Pull request by AbraaoAlves**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/elasticlunr/elasticlunr-tests.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import * as elasticlunr from 'elasticlunr';
```
Comment (RyanCavanaugh) [inline types/elasticlunr/elasticlunr-tests.ts]: Classes and functions cannot be imported using `import * as` syntax. Please fix either the test or the definition file to use a correct import form.

See also https://stackoverflow.com/questions/39415661/what-does-resolves-to-a-non-module-entity-and-cannot-be-imported-using-this

**Review by RyanCavanaugh — CHANGES_REQUESTED** [review]

File: types/elasticlunr/configuration.d.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+      *   }
+      * }
+      *
+      * then, user could search with configuration to do query-time boosting.
+      * idx.search('oracle database', {fields: {title: {boost: 2}, body: {boost: 1}}});
+      *
+      *
+      * @constructor
+      *
+      * @param {String} config user configuration
```
Comment (andy-ms) [inline types/elasticlunr/configuration.d.ts]: Don't include types in JSDoc, use TypeScript type declarations instead.

File: types/elasticlunr/configuration.d.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+      * @param {Array} fields fields of index instance
+      * @module
```
Comment (andy-ms) [inline types/elasticlunr/configuration.d.ts]: This doesn't seem like a module?

File: types/elasticlunr/configuration.d.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+      *     title: {boost: 1},
+      *     body: {boost: 0}
+      *   }
+      * }
+      *
+      * then, user could search with configuration to do query-time boosting.
+      * idx.search('oracle database', {fields: {title: {boost: 2}, body: {boost: 1}}});
+      *
+      *
+      * @constructor
```
Comment (andy-ms) [inline types/elasticlunr/configuration.d.ts]: Avoid unnecessary jsdoc tags like `@constructor`, `@module` and `@memberOf` -- this is clearly a constructor, we don't need documentation for that.

File: types/elasticlunr/document_store.d.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
+/*!
+ * elasticlunr.DocumentStore
+ * Copyright (C) @YEAR Wei Song
```
Comment (andy-ms) [inline types/elasticlunr/document_store.d.ts]: If the type definitions aren't under copyright, this comment shouldn't be included. (If they are, we'll have to review the license before merging.)

File: types/elasticlunr/document_store.d.ts [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+     * @param {String} fieldName field name
+     * @param {Integer} length field length
+     */
+    updateFieldLength(docRef?: DocRef, fieldName?: string, length?: number): void;
+    /**
+     * get field length of a document by docRef
+     *
+     * @param {Integer|String} docRef document id or reference
+     * @param {String} fieldName field name
+     * @return {Integer} field length
```
Comment (andy-ms) [inline types/elasticlunr/document_store.d.ts]: If it returns an integer, the return type should be `number`.

File: types/elasticlunr/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+// Type definitions for elasticlunr 0.9.5
+// Project: https://github.com/weixsong/elasticlunr.js
+// Definitions by: Abraão Alves <https://github.com/abraaoalves>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.5.2
```
Comment (andy-ms) [inline types/elasticlunr/index.d.ts]: If possible use a lower version. Also, this can only be `2.X`, not `2.X.Y`.

**Review by ghost — CHANGES_REQUESTED** [review]

**Pull request closed without merging**